app/controllers/catalog_controller.rb

```diff
@@ -353,29 +353,36 @@ def lazy_wps_workflow_facet
     render partial: 'lazy_wps_workflow_facet'
   end
 
+  # rubocop:disable Metrics/PerceivedComplexity
+  # rubocop:disable Metrics/CyclomaticComplexity
   def show
-    params[:id] = Druid.new(params[:id]).with_namespace
-    _deprecated_response, @document = search_service.fetch(params[:id])
+    # If showing a user version, druid will be :item_id and user_version will be :id.
+    @druid = Druid.new(params[:item_id] || params[:id]).with_namespace
+    @user_version = params.key?(:item_id) ? params[:id] : nil
+    _deprecated_response, @document = search_service.fetch(@druid)
 
-    @cocina = Repository.find_lite(params[:id], structural: false)
+    @cocina = Repository.find_lite(@druid, structural: false)
     if @cocina.instance_of?(NilModel)
       flash[:alert] =
         'Warning: this object cannot currently be represented in the Cocina model.'
     end
 
     authorize! :read, @cocina
 
-    @workflows = WorkflowService.workflows_for(druid: params[:id])
-    milestones = MilestoneService.milestones_for(druid: params[:id])
-    object_client = Dor::Services::Client.object(params[:id])
+    @workflows = WorkflowService.workflows_for(druid: @druid)
+    milestones = MilestoneService.milestones_for(druid: @druid)
+    object_client = Dor::Services::Client.object(@druid)
     versions = object_client.version.inventory
+
     user_versions = object_client.user_version.inventory
-    @milestones_presenter = MilestonesPresenter.new(milestones:, versions:, user_versions:)
+    raise ActionController::RoutingError, 'Not Found' if @user_version && user_versions.none? { |version| version.userVersion.to_s == @user_version }
+
+    @milestones_presenter = MilestonesPresenter.new(milestones:, versions:, user_versions:, druid: @druid)
```

Comment:
Take this suggestion or leave it: it looks like this method is now significantly more complex, and the milestones presenter is still on the slim side. If you were interested in re-balancing the complexity, I think you could shovel much of lines 372-377 into the presenter itself, who can take more responsibility for its own setup, given `@druid` and `@user_version`?

Reply:
That makes sense. Would you be OK if I did this in a follow-up PR? I want to unblock some of the additional work.

Reply:
@justinlittman 💯